osquery/remote/http/http_client.cpp

```diff
@@ -0,0 +1,342 @@
+        (ec.category() != boost_asio::error::system_category)) {
+      ec_ = ec;
+    }
+  });
+
+  ios_.run();
+  ios_.reset();
+
+  if (ec_) {
+    throw std::system_error(ec_.value(), adapted_category(&ec_.category()));
+  }
+
+  if (client_options_.timeout_) {
+    timer_.async_wait(
+        [=](boost_system::error_code const& ec) { timeoutHandler(ec); });
+  }
+
+  boost::beast::flat_buffer b;
+  beast_http::async_read(
+      stream, b, resp, [&](boost_system::error_code const& ec) {
+        if (client_options_.timeout_) {
+          timer_.cancel();
+        }
+        postResponseHandler(ec);
+      });
+
+  ios_.run();
+  ios_.reset();
+
+  if (ec_) {
+    throw std::system_error(ec_.value(), adapted_category(&ec_.category()));
+  }
+}
+
+Response Client::sendHTTPRequest(Request& req) {
+  if (client_options_.timeout_) {
+    timer_.expires_from_now(
+        boost::posix_time::seconds(client_options_.timeout_));
+  }
+
+  do {
+    if (req.remoteHost()) {
+      client_options_.remote_hostname_ = *req.remoteHost();
+
+      if (req.remotePort()) {
+        client_options_.remote_port_ = *req.remotePort();
+      } else if (req.protocol() && (*req.protocol()).compare("https") == 0) {
+        client_options_.remote_port_ = std::to_string(HTTPS_DEFAULT_PORT);
+      } else {
+        client_options_.remote_port_ = std::to_string(HTTP_DEFAULT_PORT);
+      }
+
+      if (req.protocol()) {
+        if ((*req.protocol()).compare("https") == 0) {
+          ssl_connection = true;
+        } else {
+          ssl_connection = false;
+        }
+      } else {
+        ssl_connection = false;
+      }
+    }
+
+    beast_http_response_parser resp;
+    createConnection();
+
+    if (ssl_connection) {
+      encryptConnection();
+    }
+
+    if (ssl_connection) {
+      sendRequest(*ssl_sock_, req, resp);
+    } else {
+      sendRequest(sock_, req, resp);
+    }
+
+    switch (resp.get().result()) {
+    case beast_http::status::moved_permanently:
+    case beast_http::status::found:
+    case beast_http::status::see_other:
+    case beast_http::status::not_modified:
+    case beast_http::status::use_proxy:
+    case beast_http::status::temporary_redirect:
+    case beast_http::status::permanent_redirect: {
+      if (!client_options_.follow_redirects_) {
+        return Response(resp.release());
+      }
+
+      std::string redir_url = Response(resp.release()).headers()["Location"];
+      if (!redir_url.size()) {
+        throw std::runtime_error(
+            "Location header missing in redirect response.");
+      }
+
+      req.uri(redir_url);
+      LOG(INFO) << "HTTP(S) request re-directed to: " << redir_url;
+      break;
+    }
+    default:
+      return Response(resp.release());
+    }
+  } while (true);
+}
+
+Response Client::put(Request& req,
+                     std::string const& body,
+                     std::string const& content_type) {
+  req.method(beast_http::verb::put);
+  req.body = body;
```

Comment:
Consider this code instead:
```
Response
Client::put(
    Request& req,
    std::string body,
    boost::beast::string_view content_type)
{
    req.method(beast_http::verb::put);
    req.body = std::move(body);
...
```

Reply:
Appears to be, we are still copying body.

Anyway I have introduced an overload function as well -
  /// HTTP post request method with rvalue reference to body param
Response Client::post(Request& req,
                      std::string&& body,
                      std::string const& content_type) {
  req.method(beast_http::verb::post);
  req.body = std::move(body);
  if (content_type.size()) {
    req.set(beast_http::field::content_type, content_type);
  }
  return sendHTTPRequest(req);
}